sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,814 @@
+and applying any properties in `props`. The memory pool will use the existing
+USM memory allocation instead of allocating from the SYCL runtime. The
+associated device and allocation kind of the memory pool is inferred from the
+`ptr`. If `ptr` is not a valid USM pointer of kind `usm::alloc::host`,
+`usm::alloc::device` or `usm::alloc::shared` or the value `size` does not match
+the size of memory allocated for `ptr` the result is undefined behavior. If
+`ptr` is freed or the memory it points to is written to whilst the memory pool
+is still alive the result is undefined behavior. The maximum size and initial
+threshold of the memory pool are set to `size`.
+
+_Throws_: An exception with the `errc::invalid` error code if `props` contains
+the `maximum_size` or `initial_threshold` properties.
+
+
+[source, c++]
+----
+~memory_pool();
+----
+
+_Effects_: If this was the last copy, signals to the SYCL runtime for the memory
+pool to be destroyed after all remaining allocations have been freed, and
+returns immediately without waiting.
+
+[source, c++]
+----
+context get_context() const;
+----
+
+_Returns_: The SYCL context associated with the memory pool.
+
+[source, c++]
+----
+device get_device() const;
+----
+
+_Returns_: The SYCL device associated with the memory pool. If the memory pool
+allocates memory of kind `usm::alloc::device` or `usm::alloc::host::shared` the
+device returned is the same device which was passed when constructing the memory
+pool. If the memory pool allocates memory of kind `usm::alloc::host` the device
+returned is the first of the devices associated with the context, which is
+associated with the memory pool.
+
+[source, c++]
+----
+usm::alloc get_alloc_kind() const;
+----
+
+_Returns_: The memory allocation kind of the memory pool.
+
+[source, c++]
+----
+size_t get_allocation_chunk_size() const;
+----
+
+_Returns_: The allocation chunk size of the device associated with the memory
+pool.
+
+[source, c++]
+----
+size_t get_max_size() const;
+----
+
+_Returns_: The maximum size of the memory pool.
+
+[source, c++]
+----
+size_t get_threshold() const;
+----
+
+_Returns_: The deallocation threshold of the memory pool.
+
+[source, c++]
+----
+size_t get_current_size() const;
+----
+
+_Returns_: The total memory allocated to the pool at the point the function is
+called.
+
+[source, c++]
+----
+size_t get_free_size() const;
+----
+
+_Returns_: The total free memory in the pool at the point the function is
+called.
+
+[source, c++]
+----
+void set_new_threshold(size_t newThreshold);
+----
+
+_Effects_: Sets the deallocation threshold of the memory pool if the value of
+`newThreshold` is larger than the current threshold. If `newThreshold` is not a
+multiple of the allocation chunk size, the new threshold is rounded down to the
+nearest multiple.
+
+_Throws_: An exception with the `errc::invalid` error code if the value of
+`newThreshold` is lower than the current threshold or larger than the maximum
+allocation size.
```

Comment:
Do you really want to throw an exception if the application attempts to lower the threshold?  What is the use case for this function?

Would it be better to name the function `increase_threshold_to` and then silently ignore the call if the threshold is already at (or greater than) `newthreshold`?